src/main/java/com/google/step/snippet/servlets/SearchServlet.java

```diff
@@ -66,41 +78,56 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
     }
 
     String query = encodeValue(param);
-    List<String> allLinks = new ArrayList<>();
+    List<Card> allCards = new ArrayList<>();
 
-    // TODO: after implementing scraping, consider changing getLink calls to a
-    // for-loop
-
-    /*
-     * Send request to retrieve card content through w3School site link from Google
-     * CSE
-     */
-    String w3Link = getLink(W3_CSE_ID, query);
-    if (w3Link != null) {
-      allLinks.add(w3Link);
-      // TODO: Call scraping function to return JSON card content
-    }
-
-    /*
-     * Send request to retrieve card content through StackOverflow site link from
-     * Google CSE
-     */
-    String stackLink = getLink(STACK_CSE_ID, query);
-    if (stackLink != null) {
-      allLinks.add(stackLink);
-      // TODO: Call stackoverflow API to return JSON card content
-    }
-
-    /*
-     * Send request to retrieve card content through GeeksForGeeks site link from
-     * Google CSE
-     */
-    String geeksLink = getLink(GEEKS_CSE_ID, query);
-    if (geeksLink != null) {
-      allLinks.add(geeksLink);
-      // TODO: Call scraping function to return JSON card content
+    ExecutorService executor = Executors.newFixedThreadPool(3);
+    List<Callable<Card>> cardCallbacks =
+        Arrays.asList(
+            () -> {
+              String w3Link = getLink(W3_CSE_ID, query);
+              if (w3Link != null) {
+                W3SchoolClient w3client = new W3SchoolClient();
+                return w3client.search(w3Link);
+              }
+              return null;
+            },
+            () -> {
+              String stackLink = getLink(STACK_CSE_ID, query);
+              if (stackLink != null) {
+                StackOverflowClient stackClient = new StackOverflowClient();
+                return stackClient.search(stackLink);
+              }
+              return null;
+            },
+            () -> {
+              String geeksLink = getLink(GEEKS_CSE_ID, query);
+              if (geeksLink != null) {
+                GeeksForGeeksClient geeksClient = new GeeksForGeeksClient();
+                return geeksClient.search(geeksLink);
+              }
```

Comment:
AFAIK most of this code for calling each of the client's respectively won't be relevant after Iris' Client interface PR since everything will be in a for loop?

Reply:
The Client interface could still be used within each Callable, but I don't think a for loop would be needed here